### PR TITLE
topsql: fix heavy loop in backoff

### LINKED ISSUE #71: topsql: not implement grpc service cause heavy loop
```
[2021/12/22 15:24:20.463 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.463 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.464 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.464 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.465 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.465 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.466 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.466 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.467 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.467 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.468 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.468 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.469 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.469 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.470 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.470 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.470 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.471 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.471 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.472 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.473 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.473 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.473 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.474 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.474 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.475 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.475 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.476 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.476 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.477 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.477 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.478 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.478 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.479 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.479 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.480 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.480 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.480 +08:00] [WARN] [scraper.go:94] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:24:20.481 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.481 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.482 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.482 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:24:20.482 +08:00] [WARN] [scraper.go:148] ["failed to receive records from stream"] [error="rpc error: code = Unimplemented desc = "]
```

### PR DESCRIPTION
**Pull request by zhongzc**
Signed-off-by: Zhenchi <zhongzc_arch@outlook.com>

### What problem does this PR solve?

Issue Number: close #71

### What is changed and how it works?

Only on calling `Recv`, grpc client emits a grpc, then throws `rpc error: code = Unimplemented`.

Previously, I expected catching `Unimplemented` on calling `Subscribe` but that's not true.

### This Patch

```
[2021/12/22 15:38:30.713 +08:00] [INFO] [manager.go:72] ["Top SQL is enabled"]
[2021/12/22 15:38:30.713 +08:00] [INFO] [scraper.go:60] ["starting to scrape top SQL from the component"] [component="{\"name\":\"tidb\",\"ip\":\"127.0.0.1\",\"port\":4000,\"status_port\":10080}"]
[2021/12/22 15:38:30.713 +08:00] [INFO] [scraper.go:60] ["starting to scrape top SQL from the component"] [component="{\"name\":\"tikv\",\"ip\":\"127.0.0.1\",\"port\":20160,\"status_port\":20180}"]
[2021/12/22 15:38:30.715 +08:00] [WARN] [scraper.go:296] ["failed to call Subscribe"] [component="{\"name\":\"tikv\",\"ip\":\"127.0.0.1\",\"port\":20160,\"status_port\":20180}"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:38:30.716 +08:00] [WARN] [scraper.go:276] ["failed to call Subscribe"] [component="{\"name\":\"tidb\",\"ip\":\"127.0.0.1\",\"port\":4000,\"status_port\":10080}"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:38:34.716 +08:00] [WARN] [scraper.go:252] ["retry to scrape component"] [component="{\"name\":\"tikv\",\"ip\":\"127.0.0.1\",\"port\":20160,\"status_port\":20180}"] [retried=1]
[2021/12/22 15:38:34.716 +08:00] [WARN] [scraper.go:252] ["retry to scrape component"] [component="{\"name\":\"tidb\",\"ip\":\"127.0.0.1\",\"port\":4000,\"status_port\":10080}"] [retried=1]
[2021/12/22 15:38:34.717 +08:00] [WARN] [scraper.go:296] ["failed to call Subscribe"] [component="{\"name\":\"tikv\",\"ip\":\"127.0.0.1\",\"port\":20160,\"status_port\":20180}"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:38:34.717 +08:00] [WARN] [scraper.go:276] ["failed to call Subscribe"] [component="{\"name\":\"tidb\",\"ip\":\"127.0.0.1\",\"port\":4000,\"status_port\":10080}"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:38:42.718 +08:00] [WARN] [scraper.go:252] ["retry to scrape component"] [component="{\"name\":\"tidb\",\"ip\":\"127.0.0.1\",\"port\":4000,\"status_port\":10080}"] [retried=2]
[2021/12/22 15:38:42.718 +08:00] [WARN] [scraper.go:252] ["retry to scrape component"] [component="{\"name\":\"tikv\",\"ip\":\"127.0.0.1\",\"port\":20160,\"status_port\":20180}"] [retried=2]
[2021/12/22 15:38:42.719 +08:00] [WARN] [scraper.go:296] ["failed to call Subscribe"] [component="{\"name\":\"tikv\",\"ip\":\"127.0.0.1\",\"port\":20160,\"status_port\":20180}"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:38:42.719 +08:00] [WARN] [scraper.go:276] ["failed to call Subscribe"] [component="{\"name\":\"tidb\",\"ip\":\"127.0.0.1\",\"port\":4000,\"status_port\":10080}"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:38:48.634 +08:00] [INFO] [manager.go:118] ["update profile target info finished"] [update-count=0]
[2021/12/22 15:38:58.720 +08:00] [WARN] [scraper.go:252] ["retry to scrape component"] [component="{\"name\":\"tidb\",\"ip\":\"127.0.0.1\",\"port\":4000,\"status_port\":10080}"] [retried=3]
[2021/12/22 15:38:58.720 +08:00] [WARN] [scraper.go:252] ["retry to scrape component"] [component="{\"name\":\"tikv\",\"ip\":\"127.0.0.1\",\"port\":20160,\"status_port\":20180}"] [retried=3]
[2021/12/22 15:38:58.721 +08:00] [WARN] [scraper.go:296] ["failed to call Subscribe"] [component="{\"name\":\"tikv\",\"ip\":\"127.0.0.1\",\"port\":20160,\"status_port\":20180}"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:38:58.722 +08:00] [WARN] [scraper.go:276] ["failed to call Subscribe"] [component="{\"name\":\"tidb\",\"ip\":\"127.0.0.1\",\"port\":4000,\"status_port\":10080}"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
[2021/12/22 15:39:30.722 +08:00] [WARN] [scraper.go:252] ["retry to scrape component"] [component="{\"name\":\"tidb\",\"ip\":\"127.0.0.1\",\"port\":4000,\"status_port\":10080}"] [retried=4]
[2021/12/22 15:39:30.722 +08:00] [WARN] [scraper.go:252] ["retry to scrape component"] [component="{\"name\":\"tikv\",\"ip\":\"127.0.0.1\",\"port\":20160,\"status_port\":20180}"] [retried=4]
[2021/12/22 15:39:30.723 +08:00] [WARN] [scraper.go:296] ["failed to call Subscribe"] [component="{\"name\":\"tikv\",\"ip\":\"127.0.0.1\",\"port\":20160,\"status_port\":20180}"] [error="rpc error: code = Unimplemented desc = "]
[2021/12/22 15:39:30.723 +08:00] [WARN] [scraper.go:276] ["failed to call Subscribe"] [component="{\"name\":\"tidb\",\"ip\":\"127.0.0.1\",\"port\":4000,\"status_port\":10080}"] [error="rpc error: code = Unimplemented desc = unknown service tipb.TopSQLPubSub"]
```

### CODE REVIEW

File: component/topsql/subscriber/scraper.go [hunk 1]
```diff
@@ -218,7 +191,43 @@ func newBackoffScrape(ctx context.Context, tlsCfg *tls.Config, address string, c
...
+	if bo.stream != nil {
+		if tidbStream, ok := bo.stream.(tipb.TopSQLPubSub_SubscribeClient); ok {
+			if record, _ := tidbStream.Recv(); record != nil {
+				return record
+			}
+		} else if tikvStream, ok := bo.stream.(resource_usage_agent.ResourceMeteringPubSub_SubscribeClient); ok {
+			if record, _ := tikvStream.Recv(); record != nil {
+				return record
+			}
+		}
```
Comment (mornyx) [inline component/topsql/subscriber/scraper.go:224]: Can use:
```go
switch bo.stream.(type) {
    case tipb.TopSQLPubSub_SubscribeClient:
        /* ... */
    case resource_usage_agent.ResourceMeteringPubSub_SubscribeClient:
        /* ... */
}
```

File: component/topsql/subscriber/scraper.go [hunk 2]
```diff
@@ -255,48 +268,47 @@ func (bo *backoffScrape) reconnect() {
 		case "tidb":
 			client := tipb.NewTopSQLPubSubClient(conn)
 			bo.client = client
-			if bo.stream, err = client.Subscribe(bo.ctx, &tipb.TopSQLSubRequest{}); err != nil {
+			stream, err := client.Subscribe(bo.ctx, &tipb.TopSQLSubRequest{})
+			if err != nil {
+				log.Warn("failed to call Subscribe", zap.Any("component", bo.component), zap.Error(err))
+				bo.retryTimes += 1
```
Comment (mornyx) [inline component/topsql/subscriber/scraper.go:274]: Just a suggestion. The current retry logic and business logic are coupled together, and each retry logic must be implemented by itself. Maybe we can implement a set of general tools in the `util` package, like:
```go
func WithRetry(times int, duration time.Duration, f func() error) error {
	var err error
	for retry := 0; retry <= times; retry++ {
		err = f()
		if err == nil {
			break
		}
		if retry < times {
			time.Sleep(duration)
		}
	}
	return err
}

func WithRetryBackoff(times int, f func() error) error {
	duration := time.Second
	var err error
	for retry := 0; retry <= times; retry++ {
		err = f()
		if err == nil {
			break
		}
		if retry < times {
			time.Sleep(duration)
			duration *= 2
		}
	}
	return err
}
```

Then we can simply do:
```go
err := WithRetry(3, time.Second, func() error {
    /* ... */
})
if err != nil {
    /* ... */
}
```

File: component/topsql/subscriber/scraper.go (hunk 1 above)
Comment (mornyx) [inline component/topsql/subscriber/scraper.go:224]: Can use:
```go
switch bo.stream.(type) {
    case tipb.TopSQLPubSub_SubscribeClient:
        /* ... */
    case resource_usage_agent.ResourceMeteringPubSub_SubscribeClient:
        /* ... */
}
```

**Comment by mornyx** [comment]
LGTM

File: component/topsql/subscriber/scraper.go (hunk 2 above)
Comment (zhongzc) [reply component/topsql/subscriber/scraper.go:268]: I will have a try.

File: component/topsql/subscriber/scraper.go (hunk 1 above)
Comment (zhongzc) [reply component/topsql/subscriber/scraper.go:224]: Nice suggestion

**Comment by zhongzc** [comment]
Comments are resolved. Can you relook again? @mornyx

File: utils/misc.go [hunk 3]
```diff
@@ -23,3 +26,57 @@ func GoWithRecovery(exec func(), recoverFn func(r interface{})) {
...
+// - f returns true. It means work is done.
+// - retry times is greater than the given times
+// - ctx is done.
+//
+// Otherwise, this function will wait a time of the given duration
+// and continue to execute f.
+//
+// The argument provided for f is the retried times.
+func WithRetry(ctx context.Context, maxTimes uint, duration time.Duration, f func(uint) bool) {
+	for retry := uint(0); retry < maxTimes; retry++ {
```
Comment (mornyx) [inline utils/misc.go:42]: It seems that we need to make it clear whether `maxTimes` means `RetryTime` or `MaxExecTime`, `MaxExecTime` needs +1 than `RetryTime`, we can explain it clearly in the comments.

File: utils/misc.go [hunk 4]
```diff
@@ -23,3 +26,57 @@ func GoWithRecovery(exec func(), recoverFn func(r interface{})) {
...
+// The given f will keep running, until:
+// - f returns true. It means work is done.
+// - retry times is greater than the given times
+// - ctx is done.
+//
+// Otherwise, this function will wait a time of the given duration
+// and continue to execute f.
+//
+// The argument provided for f is the retried times.
+func WithRetry(ctx context.Context, maxTimes uint, duration time.Duration, f func(uint) bool) {
```
Comment (mornyx) [inline utils/misc.go:41]: Can we add some unit tests for `WithRetry` and `WithRetryBackoff`? As util I think this is necessary~

File: utils/misc.go (hunk 3 above)
Comment (mornyx) [reply utils/misc.go:42]:
```go
for n := 0; n < maxExecTime; n++
```

or
```go
for n := 0; n <= retryTime; n++
```

Comment (mornyx) [reply utils/misc.go:42]: Because the first execution cannot be called "retry"

File: utils/misc.go (hunk 4 above)
Comment (zhongzc) [reply utils/misc.go:41]: ok

**Review by crazycs520 — APPROVED** [review]

**Pull request merged**